### ISSUE TITLE
x64: bug in register sizes when decoding some SSE instructions

### ISSUE THREAD

**Issue by LRGH** [bdcht/amoco#50]
Example:

``` python
>>> from amoco.arch.x64 import cpu_x64
>>> str(cpu_x64.disassemble('\xf3\x0f\x2a\xc0'))
'cvtsi2ss    xmm0, rax'
```

should return `cvtsi2ss    xmm0, eax`

Same for `cvtsi2sd`.

For these examples, it seems to be sufficient to replace `set_opdsz_64` by `set_opdsz_32` in the definition in `spec_sse.py` (and to create `set_opdsz_32` in `utils.py`).

**Comment by bdcht** [comment]
fixed by a6535e5 (v2.4.5)

**Comment by LRGH** [comment]
Not fixed:

``` python
>>> str(cpu_x64.disassemble('\xf3\x0f\x2a\xc0'))
'None'
```

because there are typos:

``` python
--- a/amoco/arch/x64/spec_sse.py
+++ b/amoco/arch/x64/spec_sse.py
@@ -402,7 +402,7 @@ def sse_sd(obj,Mod,REG,RM,data):
     REX = obj.misc['REX']
     if REX is not None:
         W,R,X,B = REX
-        if W==1: set_opdsz_64()
+        if W==1: set_opdsz_64(obj)
     op2,data = getModRM(obj,Mod,RM,data)
     op1 = getregR(obj,REG,128)
     obj.operands = [op1,op2]
@@ -536,7 +536,7 @@ def sse_sd(obj,Mod,REG,RM,data):
     REX = obj.misc['REX']
     if REX is not None:
         W,R,X,B = REX
-        if W==1: set_opdsz_64()
+        if W==1: set_opdsz_64(obj)
     op2,data = getModRM(obj,Mod,RM,data)
     op1 = getregR(obj,REG,128)
     obj.operands = [op1,op2]
```

**Comment by bdcht** [comment]
you must have been testing something else...'\xf3\x0f\x2a\xc0' has no REX prefix so it really disassemble ok (its included in test_arch_x64:test_decoder_009).
Anyway you are right about this typo...fixed by commit f357d0f.

**Comment by LRGH** [comment]
I did not test something else, but there were some side effects with the previous version.
The complete example is:

``` python
>>> from amoco.arch.x64 import cpu_x64
>>> str(cpu_x64.disassemble('\xf3\x0f\x2a\xc0'))
'cvtsi2ss    xmm0, eax'
>>> ^D

>>> from amoco.arch.x64 import cpu_x64
>>> str(cpu_x64.disassemble('\x48\x63\xd2'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "amoco/arch/core.py", line 197, in __call__
    return self(bytestring[s.mask.size/8:],**kargs)
  File "amoco/arch/core.py", line 190, in __call__
    i = s.decode(bytestring,e,i=self.__i,ival=b.ival)
  File "amoco/arch/core.py", line 437, in decode
    self.hook(obj=i,**kargs)
  File "amoco/arch/x64/spec_ia32e.py", line 629, in ia32_movsxd
    op2,data = getModRM(obj,Mod,RM,data,REX=(0,R,X,B))
NameError: global name 'R' is not defined
>>> str(cpu_x64.disassemble('\xf3\x0f\x2a\xc0'))
'None'
>>> str(cpu_x64.disassemble('\xf3\x0f\x2a\xc0'))
'cvtsi2ss    xmm0, eax'
```

It is corrected with the last commit.
